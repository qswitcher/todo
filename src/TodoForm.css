.TodoForm {
	display: flex;
	/* padding: 10px; */
	margin: 10px;
	border-radius: 6px;
	text-align: left;
}

.TodoForm input[type='text'] {
	margin-right: 10px;
	padding: 10px;
	flex-grow: 1;
	width: 50%;
	border-radius: 6px;
	border: 1px solid lightgray;
}

.TodoForm select {
	margin-right: 10px;
	width: 50px;
}

::placeholder {
	font-size: 14pt;
}

.TodoForm button {
	font-size: 14pt;
	padding: 10px 20px;
	text-align: center;
	width: 20%;
	border-radius: 6px;
	color: white;
	background-color: black;
}
