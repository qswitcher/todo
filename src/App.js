import React, { Component } from 'react';
import TodoList from './TodoList';
import TodoForm from './TodoForm';
import logo from './logo.svg';
import './App.css';

class App extends Component {
  constructor(props) {
    super(props);

    this.state = {
      todos: [
        { text: 'Walk the Dog', priority: 'High' },
        { text: 'Ride bike', priority: 'Low' },
        { text: 'Get a haircut', priority: 'Med' }
      ]
    };
  }

  addTodo = todo => {
    const todos = this.state.todos;
    // simulate an AJAX request
    setTimeout(() => {
      this.setState({ todos: todos.concat(todo) });
    }, 0);
  };

  deleteTodo = value => {
    const todos = this.state.todos;
    const index = todos.indexOf(value);
    if (index > -1) {
      this.setState({ todos: todos.filter((item, i) => i !== index) });
    }
  };

  render() {
    return (
      <div className="App">
        <header className="App-header">
          <h1 className="App-title">React Todos</h1>
        </header>
        <div className="App-body">
          <TodoForm addTodo={this.addTodo} />
          <TodoList todos={this.state.todos} deleteTodo={this.deleteTodo} />
        </div>
      </div>
    );
  }
}

export default App;
