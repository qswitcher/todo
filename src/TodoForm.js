import React, { Component } from 'react';
import './TodoForm.css';

export const priorities = ['High', 'Med', 'Low'];

class TodoForm extends Component {
	constructor(props) {
		super(props);
		this.state = { priority: priorities[2] };
	}

	onSelect = event => {
		this.setState({ priority: event.value });
	};

	render() {
		const addTodo = this.props.addTodo;
		let input;
		let select;

		return (
			<div className="TodoForm">
				<input
					type="text"
					data-testid="TodoForm-input"
					className="TodoForm-input"
					placeholder="Enter todo text"
					ref={node => {
						input = node;
					}}
				/>
				<select
					data-testid="PrioritySelect"
					value={this.state.priority}
					onChange={this.onSelect}
					name="taskPriority"
					ref={node => {
						select = node;
					}}
				>
					{priorities.map(priority => (
						<option key={priority} value={priority}>
							{priority}
						</option>
					))}
				</select>
				<button
					className="TodoForm-button"
					onClick={() => {
						addTodo({ text: input.value, priority: this.state.priority });
						input.value = '';
					}}
				>
					Add
				</button>
			</div>
		);
	}
}

export default TodoForm;
