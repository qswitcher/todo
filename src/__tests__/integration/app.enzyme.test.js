import React from 'react';
import { mount } from 'enzyme';

import App from '../../App';
import { wrap } from 'module';

import { priorities } from '../../TodoForm';

const getPriority = wr => priorities.find(p => wr.hasClass(`TodoItem--${p.toLowerCase()}`));

const getTodos = wrapper =>
  wrapper.find("[data-testid='TodoItem']").map(wr => ({
    text: wr.find("[data-testid='TodoItem-text']").text(),
    priority: getPriority(wr.find("[data-testid='TodoItem-priority']"))
  }));

let wrapper;
beforeEach(() => {
  wrapper = mount(<App />);
});

const wait = (wrapper, predicate, timeout = 10) => {
  return new Promise((resolve, reject) => {
    if (predicate(wrapper)) {
      return resolve(true);
    }
    setTimeout(() => {
      wrapper.update();
      return predicate(wrapper) ? resolve(true) : reject(new Error('Timeout expired'));
    }, timeout);
  });
};

test('entering a todo in form adds a todo', async () => {
  // enter todo text in textbox
  wrapper.find('.TodoForm-input').instance().value = 'My new todo';

  // click Add
  wrapper.find('.TodoForm-button').simulate('click');

  // wait for Todo to show up
  await wait(wrapper, w =>
    w
      .find('.TodoItem')
      .at(3)
      .exists()
  );

  // make sure form is cleared
  expect(wrapper.find('.TodoForm-input').instance().value).toEqual('');

  // make sure todo was added
  const newText = wrapper
    .find('.TodoItem')
    .at(3)
    .find('.TodoItem-text')
    .text();
  expect(newText).toEqual('My new todo');
});

test('clicking the delete button of a todo removes the todo', () => {
  expect(wrapper.find('.TodoItem').length).toEqual(3);

  // make sure 'Ride bike' is in the DOM
  expect(wrapper.contains('Ride bike')).toBeTruthy();

  // click delete button on 'Ride bike'
  wrapper
    .find('.TodoItem')
    .at(1)
    .find('button')
    .simulate('click');

  // make sure 'Ride bike' is not in the DOM
  expect(wrapper.contains('Ride bike')).toBeFalsy();
});

test('can select a prority on a todo', async () => {
  const newTodoText = 'My new todo';

  // enter new todo
  wrapper.find('.TodoForm-input').instance().value = newTodoText;

  // change the priority to high
  wrapper.find('.TodoForm select').simulate('change', { value: 'High' });

  // click add
  wrapper.find('.TodoForm-button').simulate('click');

  // wait for Todo to show up
  await wait(wrapper, w =>
    w
      .find('.TodoItem')
      .at(3)
      .exists()
  );

  const newTodo = wrapper.find('.TodoItem').at(wrapper.find('.TodoItem').length - 1);
  const newText = newTodo.find('.TodoItem-text').text();

  expect(newText).toEqual(newTodoText);

  // make sure the new todo has the right priority
  const newPriority = newTodo.find('.TodoItem-priority');
  expect(newPriority.hasClass('TodoItem--high')).toBeTruthy();

  // make sure all the todos are in the right order and have the proper priority
  const todos = getTodos(wrapper);
  expect(todos).toMatchSnapshot();
  expect(todos[todos.length - 1]).toEqual({
    text: newTodoText,
    priority: 'High'
  });
});
