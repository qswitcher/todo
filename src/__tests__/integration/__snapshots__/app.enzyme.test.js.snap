// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can select a prority on a todo 1`] = `
Array [
  Object {
    "priority": "High",
    "text": "Walk the Dog",
  },
  Object {
    "priority": "Low",
    "text": "Ride bike",
  },
  Object {
    "priority": "Med",
    "text": "Get a haircut",
  },
  Object {
    "priority": "High",
    "text": "My new todo",
  },
]
`;
